// business/models/c_vip.model.js
/***************************************************
 * 名称：c_vip实体类 
 * 描述：此代码有代码生成器生成 非技术人员请勿随便更改 
 * 版本：
 **************************************************/
var BaseModel = require('./base.model.js');

var c_vip = BaseModel.extend({
    __ClassName: 'c_vip',
    getID: function () { return this.get("ID"); },
    setID: function (v,options) {  return this.set("ID", v, options);  },
    getAD_CLIENT_ID: function () { return this.get("AD_CLIENT_ID"); },
    setAD_CLIENT_ID: function (v,options) {  return this.set("AD_CLIENT_ID", v, options);  },
    getISACTIVE: function () { return this.get("ISACTIVE"); },
    setISACTIVE: function (v,options) {  return this.set("ISACTIVE", v, options);  },
    getCARDNO: function () { return this.get("CARDNO"); },
    setCARDNO: function (v,options) {  return this.set("CARDNO", v, options);  },
    getC_VIPTYPE_ID: function () { return this.get("C_VIPTYPE_ID"); },
    setC_VIPTYPE_ID: function (v,options) {  return this.set("C_VIPTYPE_ID", v, options);  },
    getIDNO: function () { return this.get("IDNO"); },
    setIDNO: function (v,options) {  return this.set("IDNO", v, options);  },
    getVIPNAME: function () { return this.get("VIPNAME"); },
    setVIPNAME: function (v,options) {  return this.set("VIPNAME", v, options);  },
    getVIPENAME: function () { return this.get("VIPENAME"); },
    setVIPENAME: function (v,options) {  return this.set("VIPENAME", v, options);  },
    getSEX: function () { return this.get("SEX"); },
    setSEX: function (v,options) {  return this.set("SEX", v, options);  },
    getBIRTHDAY: function () { return this.get("BIRTHDAY"); },
    setBIRTHDAY: function (v,options) {  return this.set("BIRTHDAY", v, options);  },
    getVALIDDATE: function () { return this.get("VALIDDATE"); },
    setVALIDDATE: function (v,options) {  return this.set("VALIDDATE", v, options);  },
    getCREDITREMAIN: function () { return this.get("CREDITREMAIN"); },
    setCREDITREMAIN: function (v,options) {  return this.set("CREDITREMAIN", v, options);  },
    getCOUNTRY: function () { return this.get("COUNTRY"); },
    setCOUNTRY: function (v,options) {  return this.set("COUNTRY", v, options);  },
    getC_CITY_ID: function () { return this.get("C_CITY_ID"); },
    setC_CITY_ID: function (v,options) {  return this.set("C_CITY_ID", v, options);  },
    getC_OPENCARDTYPE_ID: function () { return this.get("C_OPENCARDTYPE_ID"); },
    setC_OPENCARDTYPE_ID: function (v,options) {  return this.set("C_OPENCARDTYPE_ID", v, options);  },
    getADDRESS: function () { return this.get("ADDRESS"); },
    setADDRESS: function (v,options) {  return this.set("ADDRESS", v, options);  },
    getPOST: function () { return this.get("POST"); },
    setPOST: function (v,options) {  return this.set("POST", v, options);  },
    getPHONE: function () { return this.get("PHONE"); },
    setPHONE: function (v,options) {  return this.set("PHONE", v, options);  },
    getMOBIL: function () { return this.get("MOBIL"); },
    setMOBIL: function (v,options) {  return this.set("MOBIL", v, options);  },
    getEMAIL: function () { return this.get("EMAIL"); },
    setEMAIL: function (v,options) {  return this.set("EMAIL", v, options);  },
    getC_STORE_ID: function () { return this.get("C_STORE_ID"); },
    setC_STORE_ID: function (v,options) {  return this.set("C_STORE_ID", v, options);  },
    getC_CUSTOMER_ID: function () { return this.get("C_CUSTOMER_ID"); },
    setC_CUSTOMER_ID: function (v,options) {  return this.set("C_CUSTOMER_ID", v, options);  },
    getC_CUSTOMERUP_ID: function () { return this.get("C_CUSTOMERUP_ID"); },
    setC_CUSTOMERUP_ID: function (v,options) {  return this.set("C_CUSTOMERUP_ID", v, options);  },
    getINTEGRAL: function () { return this.get("INTEGRAL"); },
    setINTEGRAL: function (v,options) {  return this.set("INTEGRAL", v, options);  },
    getPASS_WORD: function () { return this.get("PASS_WORD"); },
    setPASS_WORD: function (v,options) {  return this.set("PASS_WORD", v, options);  },
    getENTERDATE: function () { return this.get("ENTERDATE"); },
    setENTERDATE: function (v,options) {  return this.set("ENTERDATE", v, options);  },
    getVIPSTATE: function () { return this.get("VIPSTATE"); },
    setVIPSTATE: function (v,options) {  return this.set("VIPSTATE", v, options);  },
    getBEST_TIME: function () { return this.get("BEST_TIME"); },
    setBEST_TIME: function (v,options) {  return this.set("BEST_TIME", v, options);  },
    getBUILDING: function () { return this.get("BUILDING"); },
    setBUILDING: function (v,options) {  return this.set("BUILDING", v, options);  },
    getSALER: function () { return this.get("SALER"); },
    setSALER: function (v,options) {  return this.set("SALER", v, options);  },
    getSTORECARDNO: function () { return this.get("STORECARDNO"); },
    setSTORECARDNO: function (v,options) {  return this.set("STORECARDNO", v, options);  },
    getOPENCARD_STATUS: function () { return this.get("OPENCARD_STATUS"); },
    setOPENCARD_STATUS: function (v,options) {  return this.set("OPENCARD_STATUS", v, options);  },
    getOPENCARDERID: function () { return this.get("OPENCARDERID"); },
    setOPENCARDERID: function (v,options) {  return this.set("OPENCARDERID", v, options);  },
    getOPENCARDTIME: function () { return this.get("OPENCARDTIME"); },
    setOPENCARDTIME: function (v,options) {  return this.set("OPENCARDTIME", v, options);  },
    getDESCRIPTION: function () { return this.get("DESCRIPTION"); },
    setDESCRIPTION: function (v,options) {  return this.set("DESCRIPTION", v, options);  },
    getDESCRIPTION2: function () { return this.get("DESCRIPTION2"); },
    setDESCRIPTION2: function (v,options) {  return this.set("DESCRIPTION2", v, options);  },
    getC_SERVICEAREA_ID: function () { return this.get("C_SERVICEAREA_ID"); },
    setC_SERVICEAREA_ID: function (v,options) {  return this.set("C_SERVICEAREA_ID", v, options);  },
    getAGE: function () { return this.get("AGE"); },
    setAGE: function (v,options) {  return this.set("AGE", v, options);  },
    getOPENCARDDATE: function () { return this.get("OPENCARDDATE"); },
    setOPENCARDDATE: function (v,options) {  return this.set("OPENCARDDATE", v, options);  },
    getHR_EMPLOYEE_ID: function () { return this.get("HR_EMPLOYEE_ID"); },
    setHR_EMPLOYEE_ID: function (v,options) {  return this.set("HR_EMPLOYEE_ID", v, options);  }
});

module.exports = c_vip;
