// business/validates/c_vip.validate.js
/***************************************************
 * 名称：c_vip实体类对应的校验器 
 * 描述：此代码有代码生成器生成 非技术人员请勿随便更改 
 * 版本：
 **************************************************/

module.exports = {
        ID: {
    },
    AD_CLIENT_ID: {
    },
    ISACTIVE: {
    },
    CARDNO: {
    },
    C_VIPTYPE_ID: {
    },
    IDNO: {
    },
    VIPNAME: {
    },
    VIPENAME: {
    },
    SEX: {
    },
    BIRTHDAY: {
    },
    VALIDDATE: {
    },
    CREDITREMAIN: {
    },
    COUNTRY: {
    },
    C_CITY_ID: {
    },
    C_OPENCARDTYPE_ID: {
    },
    ADDRESS: {
    },
    POST: {
    },
    PHONE: {
    },
    MOBIL: {
    },
    EMAIL: {
    },
    C_STORE_ID: {
    },
    C_CUSTOMER_ID: {
    },
    C_CUSTOMERUP_ID: {
    },
    INTEGRAL: {
    },
    PASS_WORD: {
    },
    ENTERDATE: {
    },
    VIPSTATE: {
    },
    BEST_TIME: {
    },
    BUILDING: {
    },
    SALER: {
    },
    STORECARDNO: {
    },
    OPENCARD_STATUS: {
    },
    OPENCARDERID: {
    },
    OPENCARDTIME: {
    },
    DESCRIPTION: {
    },
    DESCRIPTION2: {
    },
    C_SERVICEAREA_ID: {
    },
    AGE: {
    },
    OPENCARDDATE: {
    },
    HR_EMPLOYEE_ID: {
    }
}
