// framework/lib/check.js
/**************************************************
 * 名称：通用类型检测工具
 * 日期：2014-11-20
 * 描述：准确检测js对象的类型
 * 作者：Beven
 **************************************************/
var CheckQuestion = function () { }

module.exports = new CheckQuestion();

/*
 * 名称：检测指定变量是否为指定类型
 * 参数(v)：值
 */
CheckQuestion.prototype.isType = function (obj, type) {
    return Object.prototype.toString.apply(obj) == "[object " + type + "]";
}
/*
 * 名称：检测当前对象是否为数组
 * 参数(v)：值
 */
CheckQuestion.prototype.isArray = function (obj) {
    return this.isType(obj, "Array");
}
/*
 * 名称：检测当前对象是否为Object
 * 参数(v)：值
 */
CheckQuestion.prototype.isObject = function (obj) {
    return obj != null && this.isType(obj, "Object");
}
/*
 * 名称：检测当前对象是否为Date
 * 参数(v)：值
 */
CheckQuestion.prototype.isDate = function (obj) {
    return this.isType(obj, "Date");
}
/*
 * 名称：检测当前对象是否为String
 * 参数(v)：值
 */
CheckQuestion.prototype.isString = function (obj) {
    return this.isType(obj, "String");
}
/*
 * 名称：检测当前对象是否为Number
 * 参数(v)：值
 */
CheckQuestion.prototype.isNumber = function (obj) {
    return this.isType(obj, "Number");
}
/*
 * 名称：检测当前对象是否为Boolean
 * 参数(v)：值
 */
CheckQuestion.prototype.isBoolean = function (obj) {
    return this.isType(obj, "Boolean");
}
/*
 * 名称：检测当前对象是否为Function
 * 参数(v)：值
 */
CheckQuestion.prototype.isFunction = function (obj) {
    return this.isType(obj, "Function");
}
/*
 * 名称：检测当前对象指定属性是否存在 如果不存在则进行使用默认值进行初始化
 * 参数(v)：值
 */
CheckQuestion.prototype.noInObj = function (obj, p, dv) {
    var v = obj?obj[p]: dv;
    if (v == null) {
        obj[p] = v = this.isFunction(dv)? dv():dv;
    }
    return v;
}
/*
 * 名称：检测传入值是否为null或者undefined或者为''空字符串
 * 参数(v)：值
 */
CheckQuestion.prototype.isEmpty = function (v) {
    if (v == null || v == '' || v == undefined) {
        return true;
    } else {
        return false;
    }
}
