// framework/lib/log.js
/*************************************************************
 * 基于log4js-node的日志
 * 
 * 
 * **************************************************************/
var log4js = require('log4js-node');
var check = require('./check.js');
var path = require('path');
var io = require('fs-extra');

//全局日志对象
var logger = log4js.getLogger();

var LogUtil = function (options) { init(options); }

exports = LogUtil;

//输入调试日志
LogUtil.prototype.Debug = function (message) {
    logger.debug(message);
}

//输出错误日志
LogUtil.prototype.Error = function (err) {
    logger.error(err);
}

//输出信息日志
LogUtil.prototype.Info = function (infoMessage) {
    logger.info(infoMessage);
}

//初始化log4js配置
function init(options) {
    initAppenders(options);
    log4js.configure(options);
}

//初始化log4js配置
function initAppenders(options) {
    if (options) {
        var appenders = options.appenders;
        if (check.isArray(appenders)) {
            var appender = null;
            for (var i = 0, k = appenders.length; i < k; i++) {
                appender = appenders[i];
                initAppender(appender);
            }
        }
    }
}

//初始化log4js appender对象 并且自动创建对应的目录
function initAppender(appender) {
    if (appender) {
        var dir = path.basename(appender.filename);
        io.ensureDir(dir);
    }
}
