// framework/codegen/base.gen.js
/***************************************************
 * 名称：Webpos2.0代码生成工具
 * 描述：主要用于生成Model,ModelCollection,View以及Validator配置等文件
 **************************************************/
var SQLiteAccessor = require('Accessor').SQLiteAccessor;
var PathUtil = require('path');
var StrUtil = require('../lib/string.js');
var TypeQues = require('../lib/check.js');
var GenContext = require('./context.js');
var BaseCodeGener = function () { this.Init(); }

var GENERS = [];

BaseCodeGener.prototype.CodeGen = function (table, callback) {
    var context = new GenContext(table, callback);
    context.BaseDIR = PathUtil.join(__dirname, '..', '..', 'business');
    this.ReadyColumns(context);
}
//初始化生成器列表
BaseCodeGener.prototype.Init = function () {
    GENERS.length = 0;
    GENERS.push(require('./model.gen.js'));
    GENERS.push(require('./validate.gen.js'));
    GENERS.push(require('./collection.gen.js'));
    GENERS.push(require('./view.gen.js'));
}
//获取指定表的元数据信息
BaseCodeGener.prototype.ReadyColumns = function (context) {
    var accessor = new SQLiteAccessor(context.TableName);
    var self = this;
    accessor.query(StrUtil.Format("PRAGMA table_info(`{0}`);", context.TableName), function (err, data) {
        self.ReadyColumnsFinal(err, data, context);
    });
}
//获取指定表的元数据完毕
BaseCodeGener.prototype.ReadyColumnsFinal = function (err, data, context) {
    if (err) {
        return context.Final(err, context);
    }
    context.Columns = data;
    this.Gening(context);
}
//开始启动生成
BaseCodeGener.prototype.Gening = function (context) {
    this.Call(this.OnStaring, context);
    var geners = GENERS;
    for (var i = 0, k = geners.length; i < k; i++) {
        geners[i].CodeGen(context);
    }
    this.Gened(context);
}
//生成完毕
BaseCodeGener.prototype.Gened = function (context) {
    this.Call(this.OnGened, context);
    context.Final(context.Error, context);
}
//事件列表
//开始生成事件
BaseCodeGener.prototype.OnStaring = function (context) {
    console.log("start gen...");
}
//生成完毕事件
BaseCodeGener.prototype.OnGened = function (context) {
    console.log("gen all finish!");
}
//安全执行传入函数
BaseCodeGener.prototype.Call = function (fn, arg1, arg2) {
    if (TypeQues.isFunction(fn)) {
        var args = Array.prototype.splice.call(arguments, 1);
        return fn.apply(this, args);
    }
}

module.exports = new BaseCodeGener();
