// framework/codegen/context.js
/***************************************************
 * 名称：Webpos2.0代码生成上下文
 * 日期：2014-12-15
 * 描述：无
 * 作者：Beven
 **************************************************/
var typeQues = require('../lib/check.js');

var CodeGenContext = function (table) { Init.apply(this, arguments); }
module.exports = CodeGenContext;
//表名
CodeGenContext.prototype.TableName = "";
//字段数据信息
CodeGenContext.prototype.Columns = [];
//执行完毕的回调函数
CodeGenContext.prototype.CallBack = null;
//执行异常信息
CodeGenContext.prototype.Error = null;
//生成文件目标目录
CodeGenContext.prototype.BaseDIR = "";
//本次生成版本
CodeGenContext.prototype.Version = "";

/*
 * 名称：生成完毕的标记函数 调用此函数会调用已经注册回调函数
 * 参数(err)：生成结果是否出现错误的错误消息
 * 参数(result)：生成结果参数
 */
CodeGenContext.prototype.Final = function (err, result) {
    if (typeQues.isFunction(this.CallBack)) {
        this.CallBack(err, result);
    }
}

/*
 * 名称：初始化生成上下文参数
 * 参数(table)：表名
 * 参数(callback)：生成完毕即调用Final函数时被调用的回调函数
 */
function Init(table, callback) {
    this.TableName = table;
    this.CallBack = callback;
}
