// framework/framework.js
/******************************************************************************
 *  webpos 2.0 框架主类 
 *  相关方法以及类可以使用此模块进行访问与使用
 *  
 * *****************************************************************************/
var path = require('path');
var util = require('util');
var log = require('./lib/log.js');

var Framework = function () { }

module.exports = new Framework();

function getReleativeAppDomainBaseDIR() {
    return path.join(__dirname, '..', '..', './');
}

//全局配置 优先级最高 位置尽量靠前
Framework.prototype.Configuration = require('../.././config/configuration.js');

//处理工具
Framework.prototype.Util = require('./lib/util.js');

//日志记录工具
Framework.prototype.Logger = new log(Framework.prototype.Configuration.log4jsCfg, getReleativeAppDomainBaseDIR());

//检测工具
Framework.prototype.CheckQuestion = require('./lib/check.js');

//全局消息
Framework.prototype.InfoMessageStrings = require('./lib/locale/localize.js');

//异常创建者
Framework.prototype.ErrorBuilder = require('./lib/exception.js');

//字符串操作辅助工具
Framework.prototype.String = require('./lib/string.js');

//时间处理工具
Framework.prototype.Date = require('./lib/date.js');

//增量数据同步
Framework.prototype.SyncDataIncrement = require('./lib/incremental.js');

//数据接口
Framework.prototype.DataAPI = require('./lib/data.api.js');

//模块校验类
Framework.prototype.ModuleValidate = require('./lib/module.validate.js');

//代码生成
Framework.prototype.CodeGener = require('./codegen/base.gen.js');

//获取程序运行根目录
Framework.prototype.GetAppDomainBaseDIR = function () {
    return this.CheckQuestion.noInObj(root, 'baseDIR', getReleativeAppDomainBaseDIR);
}

//安全指定传入函数，如果传入的不是函数类型 不引发异常 调用方式  this.Call(函数,参数1,参数2,....,参数N)  相当于：函数(参数1,参数2...参数N);
Framework.prototype.Call = function (fn, arg1, argN) {
    if (this.CheckQuestion.isFunction(fn)) {
        var args = Array.prototype.slice.call(arguments, 1);
        return fn.apply(null, args);
    }
}

//安全指定传入函数，如果传入的不是函数类型 不引发异常   this.Apply(函数,数组参数) 相当于: 函数(args[0],args[1],...args[N]);
Framework.prototype.Apply = function (fn, args) {
    if (this.CheckQuestion.isFunction(fn)) {
        return fn.apply(null, args);
    }
}

//加载指定模块的扩展块  name:模块名称 例如:backbone 那么会require('./extensions/backbone.extension.js');
Framework.prototype.Extension = function (name) {
    return require('./extensions/' + name + '.extension.js');
}
